### PR TITLE
[part 2/n] Move kmod and BPF engines to vtables

### PR DESCRIPTION
**Pull request by gnosek**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

/kind cleanup

/kind design

> /kind documentation

> /kind failing-test

> /kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

/area libscap-engine-bpf

> /area libscap-engine-gvisor

/area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-udig

/area libscap

> /area libsinsp

> /area tests

> /area proposals

**What this PR does / why we need it**:

This PR continues the work started in https://github.com/falcosecurity/libs/pull/213 and makes kmod and BPF independent scap engines. It should be a huge no-op that makes future maintenance much easier.

**Special notes for your reviewer**:

I tried (not sure if I succeeded) to make the individual commits reviewable, so there's a bunch of no-op changes in there that do nothing except make future diffs smaller. Hopefully this splits the patches into two categories:
* large but boring (just scroll past)
* smaller and more interesting (take a closer look)

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: userspace/libscap/scap_procs.c [hunk 1]
```diff
@@ -1714,24 +1670,24 @@ int32_t scap_check_suppressed(scap_t *handle, scap_evt *pevent, bool *suppressed
...
+int32_t scap_procfs_get_threadlist(struct scap_engine_handle engine, struct ppm_proclist_info **procinfo_p)
 {
-	struct ppm_proclist_info *res = NULL;
 #ifdef __linux__
 	DIR *dir_p = NULL;
 	DIR *taskdir_p = NULL;
 	FILE *fp = NULL;
 	struct dirent *dir_entry_p;
 	char procdirname[SCAP_MAX_PATH_SIZE];
+	struct kmod_engine* kmod_engine = engine.m_handle;
```
Comment (FedeDP) [inline userspace/libscap/scap_procs.c:1681]: If i am not wrong, udig engine too has `scap_procfs_get_threadlist` in its vtable :)

File: userspace/libscap/scap_vtable.h [hunk 2]
```diff
@@ -223,6 +223,50 @@ struct scap_vtable {
...
+	/**
+	 * @brief get the list of all threads in the system, with their cpu usage
+	 * @param engine wraps the pointer to the engine-specific handle
+	 * @param procinfo_p pointer to pointer to the resulting list
+	 * @return SCAP_SUCCESS or a failure code
+	 *
+	 * `procinfo_p` must not be NULL, but `*procinfo_p` may be; the returned
+	 * list will be (re)allocated on demand
+	 */
+	int32_t (*get_threadlist)(struct scap_engine_handle engine, struct ppm_proclist_info **procinfo_p);
```
Comment (FedeDP) [inline userspace/libscap/scap_vtable.h:236]: Just a side question: is there any actual check that when a vtable is defined, all of its fields are actually non-NULL?
Or, we could do that if they're NULL, scap automatically sets the no-op for you. 
This way no-op vtables get much shorter (just a bunch of NULL and a name).

**Review by FedeDP — COMMENTED** [review]
Initial review; still missing quite a bit of things though.

File: userspace/libscap/engine/bpf/bpf_public.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+limitations under the License.
+*/
+
+#pragma once
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+const char *scap_get_bpf_probe_from_env();
```
[the remark this answers is not in the document]
Comment (gnosek) [reply userspace/libscap/engine/bpf/bpf_public.h:23]: We use it in sysdig-the-command-line-tool. I'd be very happy to remove it from both places but this PR series is already huge ;)

Some Day™

File: userspace/libscap/scap_vtable.h (hunk 2 above)
Comment (gnosek) [reply userspace/libscap/scap_vtable.h:236]: It's C, so there are no checks and we're in full yolo mode. Somewhat more seriously, we'd probably want to skip the check for .next() anyway. Also, all the vtables are const and we can't modify them at runtime (we can copy them and modify the copy if we want).

The check for a full vtable would have to be manually written too (can't introspect the struct automatically).

All in all, it's not the highest priority for me right now, but I'm certainly not going to oppose a follow up PR that handles this better :)

File: userspace/libscap/engine/bpf/bpf_public.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (FedeDP) [reply userspace/libscap/engine/bpf/bpf_public.h:23]: Some Day :star_struck:

File: userspace/libscap/scap_vtable.h (hunk 2 above)
Comment (FedeDP) [reply userspace/libscap/scap_vtable.h:236]: I will give this a look once everything is in place; thanks!

File: userspace/libscap/engine/bpf/scap_bpf.c [hunk 4]
```diff
@@ -80,6 +86,63 @@ struct bpf_map_data {
...
+
+	snprintf(buf, SCAP_MAX_PATH_SIZE, "%s/%s", home, SCAP_PROBE_BPF_FILEPATH);
+	return buf;
+}
+
+static bool match(scap_open_args* open_args)
+{
+	char bpf_probe_buf[SCAP_MAX_PATH_SIZE];
+
+	open_args->bpf_probe = resolve_bpf_probe(open_args->bpf_probe, bpf_probe_buf);
```
[the remark this answers is not in the document]
Comment (gnosek) [reply userspace/libscap/engine/bpf/scap_bpf.c:127]: Yes, with a matching comment (in scap_open_live_int IIRC) indicating that it's a very bad idea. After I realized it's pointless, as well as dangerous, I've cleaned this up locally, haven't rebased & pushed yet. Good catch :)

**Review by Andreagit97 — COMMENTED** [review]
That's huge work @gnosek! Thank you!

File: userspace/libscap/scap_procs.c (hunk 1 above)
Comment (gnosek) [reply userspace/libscap/scap_procs.c]: Done

File: userspace/libscap/engine/bpf/scap_bpf.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (gnosek) [reply userspace/libscap/engine/bpf/scap_bpf.c]: Done

**Review by Andreagit97 — APPROVED** [review]
/approve

**Review by FedeDP — APPROVED** [review]
/approve

**Review by jasondellaluce — COMMENTED** [review]
This is huge work! LGTM

**Pull request merged**